### PR TITLE
Add Apache Pulsar support

### LINKED ISSUE #21553: Apache Pulsar support
##### **Overview of the feature request**

Add an option "Asynchronous messages using Apache Pulsar" in other technologies.

##### **Motivation for or Use Case**

Apache Pulsar has a lot of momentum.
[Spring-Pulsar 0.2.0](https://spring.io/blog/2023/03/24/spring-for-apache-pulsar-0-2-0-available-now) was just released.
I'm working on a PR for this.

##### **Related issues or PR**

- [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by cbornet**
Fix #21553 

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs
Comment (mshima) [inline generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs:142]: Extract to its own file like [RedisTestContainersSpringContextCustomizerFactory.java.ejs ](https://github.com/jhipster/generator-jhipster/blob/main/generators/spring-cache/templates/src/test/java/package/config/RedisTestContainersSpringContextCustomizerFactory.java.ejs).
And inject like

https://github.com/jhipster/generator-jhipster/blob/f3d8124aad13ce33cbc4e8d48f3fd4e8e46da235/generators/spring-cache/generator.mts#L56-L59

File: generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs
Comment (mshima) [inline generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs:47]:
```suggestion
```

File: generators/server/templates/build.gradle.ejs
Comment (mshima) [inline generators/server/templates/build.gradle.ejs:450]: Add to it's own pulsar.gradle like [cache.gradle](https://github.com/jhipster/generator-jhipster/blob/main/generators/spring-cache/templates/gradle/cache.gradle.ejs)
And inject like
https://github.com/jhipster/generator-jhipster/blob/f448a5a3fbe27967df92495080b2391140132648/generators/spring-cache/generator.mts#L62-L66

File: generators/server/templates/build.gradle.ejs
Comment (mshima) [inline generators/server/templates/build.gradle.ejs:327]: Add those dependencies to pulsar.gradle too.
```suggestion
<%_ if (databaseTypeMongodb || databaseTypeCouchbase || databaseTypeNeo4j || databaseTypeCassandra || searchEngineElasticsearch || messageBrokerKafka || databaseTypeSql) { _%>
```

**Review by mshima — CHANGES_REQUESTED** [review]
Replace every `messageBrokerKafka || messageBrokerPular` with `messageBrokerAny`.

File: generators/server/templates/build.gradle.ejs
Comment (cbornet) [reply generators/server/templates/build.gradle.ejs:450]: We don't do that for other deps (kafka, swagger, ...).
Shouldn't this be done in another PR and keep consistency for the moment ?

File: generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs
Comment (cbornet) [reply generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs:47]: The other beans don't have these skipped lines.
It should be kept consistent.

Comment (mshima) [reply generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs:47]: Related to https://github.com/jhipster/generator-jhipster/pull/21563#discussion_r1149131359.

File: generators/server/templates/build.gradle.ejs
Comment (mshima) [reply generators/server/templates/build.gradle.ejs:450]: I've started to split recently when working on a specific generator.
Otherwise the PR can grow too much.
Kafka is done at https://github.com/jhipster/generator-jhipster/pull/21577.

**Comment by mshima** [comment]
Should add pulsar to an integration test too.
Example: add to every app at https://github.com/jhipster/generator-jhipster/blob/main/test-integration/jdl-samples/mf-ng-eureka-jwt-psql-ehcache/blog-store.jdl.
This will make sure generated application stack with reactive/imperative, gradle/maven integration tests passes.

File: generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs
Comment (cbornet) [reply generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs:47]: Oh, sorry. I misunderstood the change.

File: generators/server/templates/build.gradle.ejs
Comment (cbornet) [reply generators/server/templates/build.gradle.ejs:450]: OK. I'll do it then.

**Comment by cbornet** [comment]
> Should add pulsar to an integration test too.

done

File: generators/server/templates/build.gradle.ejs
Comment (cbornet) [reply generators/server/templates/build.gradle.ejs]: done

File: generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs
Comment (cbornet) [reply generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs]: done

File: generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs
Comment (cbornet) [reply generators/server/templates/src/test/java/package/config/TestContainersSpringContextCustomizerFactory.java.ejs]: done

File: generators/languages/templates/src/main/webapp/i18n/cs/health.json.ejs
Comment (mshima) [inline generators/languages/templates/src/main/webapp/i18n/cs/health.json.ejs:32]:
```suggestion
<%_ if (messageBrokerAny) { _%>
```

File: generators/pulsar/generator.mjs
Comment (mshima) [inline generators/pulsar/generator.mjs:52]: Files customizations were lost in rebase, we should use api instead:
```suggestion
      customizeSpring({ source, application }) {
        source.addIntegrationTestAnnotation?.({ package: `${application.packageName}.config`, annotation: 'EmbeddedPulsar' });
```
Re-add logs?
https://github.com/jhipster/generator-jhipster/blob/32f8f609d2a028c9c6e4e5cee17397537a6db1ea/generators/kafka/generator.mts#L48-L51

File: generators/pulsar/files.mjs
Comment (mshima) [inline generators/pulsar/files.mjs:40]: I think we should move to a package.
```suggestion
        'broker/PulsarIT.java',
```

**Review by mshima — CHANGES_REQUESTED** [review]

**Comment by mshima** [comment]
The generator changes are ok 👍.
Need to fix the generated application tests.
The error is on reactive only?
You can disable pulsar at gateway and ms reactive to test.

@DanielFran should we unify Kafka and pulsar at one generator `spring-cloud-stream` on another PR? Both implementation looks compatible with the dependency change.

**Comment by cbornet** [comment]
> The error is on reactive only?

Yes. It looks like the Pulsar client makes a blocking call to load a JAR.
I've added an `allowBlockingCallsInside` to blockhound for that.

**Comment by cbornet** [comment]
> @DanielFran should we unify Kafka and pulsar at one generator spring-cloud-stream on another PR? Both implementation looks compatible with the dependency change.

Yes. Actually I think the Kafka generator generates too much things.
Here for Pulsar, I only generate a test to check that the integration with Spring Cloud Stream and testcontainers is correct.
I think the code could be exactly the same for Kafka.

**Comment by DanielFran** [comment]
> > @DanielFran should we unify Kafka and pulsar at one generator spring-cloud-stream on another PR? Both implementation looks compatible with the dependency change.
> 
> Yes. Actually I think the Kafka generator generates too much things. Here for Pulsar, I only generate a test to check that the integration with Spring Cloud Stream and testcontainers is correct. I think the code could be exactly the same for Kafka.

OK, so better to merge this when validated and we can then merge both generators into one unique.

**Comment by mshima** [comment]
> Yes. Actually I think the Kafka generator generates too much things. 

There is a legacy rest implementation with registration IMO we should keep or make optional, and a supplier for test purpose that can be moved to test.

**Review by mshima — APPROVED** [review]
Tests still failing, once fixed can be merged.

**Comment by cbornet** [comment]
Yes I've seen that 😢 . Hard to diagnose as it doesn't fail on my machine...

**Comment by cbornet** [comment]
Oh, I screwed up with the blockhound config. Let me try again.

File: generators/pulsar/templates/gradle/pulsar.gradle.ejs
Comment (mshima) [inline generators/pulsar/templates/gradle/pulsar.gradle.ejs:4]: Conflict with sb 3.0.2?
Please add a description.

**Review by DanielFran — APPROVED** [review]

File: generators/pulsar/templates/gradle/pulsar.gradle.ejs
Comment (cbornet) [reply generators/pulsar/templates/gradle/pulsar.gradle.ejs:4]: Yes. It pulls SB 3.0.5 and undertow autoconfig is broken. Between patch versions...

Comment (onobc) [reply generators/pulsar/templates/gradle/pulsar.gradle.ejs:4]: `3.0.3` moved to UT`2.3.4.Final` as per https://github.com/spring-projects/spring-boot/issues/34273

The UT dependency has been steady in 3.0.3+
https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-dependencies/3.0.(3|4|5)

- `3.0.3 -> 2.3.4.Final`
- `3.0.4 -> 2.3.4.Final`
- `3.0.5 -> 2.3.4.Final`


The change in SB `3.0.3+` does require this change in Undertow in https://github.com/undertow-io/undertow/commit/e2a1f3ad39d96312f3befbe302629f4a4c2b5cdb

Aka the newly added `DECODE_SLASH`.

@cbornet can you verify what version of UT is being used? I wonder if its older (w/o the newly added field).

Comment (onobc) [reply generators/pulsar/templates/gradle/pulsar.gradle.ejs:4]: If Jhipster is on `3.0.2` I could see how this would be an issue though.

Basically if you go from SB `3.0.2` to SB `3.0.3+` you have to upgrade Undertow as well. 😿 .

Comment (cbornet) [reply generators/pulsar/templates/gradle/pulsar.gradle.ejs:4]: Except there was a change in SB to use that field. So if the undertow being pulled in the old one...

**Comment by mshima** [comment]
Closing reopen to test with SB 3.0.5

File: generators/pulsar/templates/gradle/pulsar.gradle.ejs
Comment (onobc) [reply generators/pulsar/templates/gradle/pulsar.gradle.ejs:4]: > So if the undertow being pulled in the old one...

Agreed - w/o the update to undertow we get the exception.

**Pull request merged**
